### DISCUSSION TITLE
zipped synctex

### DISCUSSION THREAD

**Discussion by honekamp** [sumatrapdfreader/sumatrapdf#2640] [General]
In older versions of Sumatra PDF, it was possible to have a zipped synctex file and still backwards and forward search would work just fine. Starting 3.4.x, the support for a zipped synctex file seems to have ceased. I have to unzip the synctex file in order to do a forward or backward search. Has this change been applied intentionally? Is there any way to work around the change?

**Comment by GitHubRulesOK** [comment]
Probably not intentional, last time I looked it was working 
Argh! it was not zipped can you provide a good mini sample (one page with few simple variations will do)
mwe.pdf mwe.tex and mwe.synctex.gz (you may need to lump all in a conventional zip to paste in a comment 🙂 )

@kjk 
You may have changed synctex gzip library handling inadvertently?

**Reply by honekamp** [reply]
I‘m in the process of creating a simple example. It may take a little while, I just want to make sure that I get it right.

**Comment by honekamp** [comment]
[Test.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/8805799/Test.zip)

I have created a small example that hopefully exemplifies the issue. As long as I don‘t unzip the synctex file, neither forward nor backward search work for me. After unzipping, search works again. Just in case the information is relevant, I‘m using Sublime Text with LaTeX Tools for authoring LaTeX sources.

**Reply by GitHubRulesOK** [reply]
Thanks
 editor should not matter it is the command part of syntax with `pdflatex -1 synctex` or something (bad memory) where there are many (15?) variations 0=uncompressed

**Reply by GitHubRulesOK** [reply]
I should have asked which version of SumatraPDF ?

**Reply by honekamp** [reply]
I‘m using 3.4.3, the latest as far as I know.

**Comment by GitHubRulesOK** [comment]
Hmmmm
@kjk 
It seems to work as expected in 3.5.15009

![image](https://user-images.githubusercontent.com/12754913/171215976-96ac402d-6cc6-4f55-9cd8-0b4b94c4d49d.png)

the file is compressed (large at 32kb but 7 pages and last paragraph starts at `mwe.tex 154` = Duis...... which tallies too
<strike>I thus have to suggest it is the editor configuration of SumatraPDF
 and recollect there were others with similar recent question using VScode due to a plugin change.</strike>

**Comment by GitHubRulesOK** [comment]
Ahh
@kjk 3.4.3 seems to be the culprit 3.5... OK 

3.4.3 say no synchronization open issue https://github.com/sumatrapdfreader/sumatrapdf/issues/2642#
![image](https://user-images.githubusercontent.com/12754913/171224439-dcf979b5-d207-4c56-b369-0444b55dfb81.png)

**Reply by honekamp** [reply]
Thanks a lot for the investigation!

**Comment by kjk** [comment]
@honekamp thanks for the sample.

However, it works for me in 3.4.3 (64 -bit, installer).

What I did:
* extracted the `Test.zip` file
* I opened `mwe.pdf`
* I double-clicked on a word
* I have notepad++ installed so sumatra auto-detected it and launched notepad++ with `mwe.tex` positioned at the right line

This is all using `mwe.synctex.gz`.

So what happens in your setup when after you double-click on a word? Nothing? An error message?

Are you using a portable version or installer?

**Comment by kjk** [comment]
I've tested `mwe.pdf` with portable version of 3.1.2, 3.2, 3.3 and 3.4.3, 64-bit, and they all work with `mwe.synctex.gz` and notepad++

The auto-detected inverse-search command is: `"C:\Program Files\Notepad++\notepad++.exe" -n%l "%f"`

So maybe the issue with your setup is inverse-search command. What is it?

**Reply by honekamp** [reply]
It is strange that the sample does not exhibit the described behavior, but the whole issue is strange. 

Maybe I should mention that I‘m a daily user of Sumatra PDF (installer) for a long time (i.e. since before it reached 1.0). My setup is very stable, I‘ve used Sublime Text as my LaTeX editor all these years (with a short excursion to VS Code in between, because new and shiny). Both forward (from Sublime to Sumatra PDF) and inverse search (from Sumatra PDF to Sublime) worked reliably day by day all these years until I switched from Sumatra PDF 3.3.3 to 3.4.3 this week. 

For forward search, Sumatra PDF 3.4.3 gives me an error message that the „Synchronization file cannot be opened“. The error message for inverse search is „No synchronization info at this position“.

I tried the 3.5 pre-release build, but that one crashes during the forward search in my real projects. If I install 3.3.3, everything is back to normal, both search directions work with the zipped synctex file.

After reading the error message, I tried to find a hint how to work around the issue. After some experimentation, I found out that the search in both directions started working again in 3.4.3 as soon as I unzipped the zipped synctex file. Unfortunately, I cannot provide a sample of the real projects I‘m working on because these are very complex and the build requires some tooling effort. The documents are typically slightly below or larger than a 1000 pages.

Of course, a possible work-around for me could be to just tweak the build workflow such that the synctex file is left uncompressed.

**Reply by kjk** [reply]
The crash you saw with pre-release should be fixed.

I've added logging to pre-release around synctex handling. The build should be ready in about 1 hr.

When it is, could you test with pre-release (15032 or later)?

We narrowed it down to Sumatra failing to parse synctex.gz, most likely because it fails to open it.

I've added logging around opening / parsing .synctex.gz files.

When the issue happens with pre-release, you can use `Ctrl + K` for command palette and then `show log` to open the log in notepad. Then you could post the logs here.

My only guess so far that it's a timing issue.  

Are your projects big? How big is `.synctext.gz` file? 

Maybe when Sumatra is trying to load `synctext.gz` it isn't yet fully generated and that's why it fails to parse. The logs would provide more info.

**Comment by honekamp** [comment]
[sumatra-log.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/8811700/sumatra-log.zip)

I have uploaded the log file obtained from build 15033 for the case of a zipped synctex file. Forward search does not work, and the error message is „No synchronization file found“. If I unzip the synctex file and try to do a forward search, then build 15033 simply crashes. 

Please note that I have deleted several thousand lines of file states in the log file because these might contain sensitive information.

BTW, the syntex.gz file for the project I ran the experiment on is a about 12 MB.

**Reply by kjk** [reply]
Please retry with pre-release 15039 or later and submit the logs.

I've added logging of synctex code so that we can pin point what goes wrong there.

Also, I've added a retry with 1 sec so if my theory about timing issue is right, this could fix it.

**Reply by honekamp** [reply]
I tried with build 15051. Initially, I initially made a mistake because the project I was taking for a text had an non-zipped synctex file. The log obtained from that experiment is probably worthless. I deleted the non-zipped synctex file such that only the zipped version remains.

A forward search yields the error message "No synchronization file found" and **does not create a log file**. Inverse sync does not work and the error message "No synchronization info at this position" is created.

**Reply by GitHubRulesOK** [reply]
@honekamp 
KJK is the developer thus able to see where the issue may be however, I was at one time able to emulate your problem.
I have to agree it is likely to be that when you double click, SumatraPDF needs to expand a compressed file so as to read, the encoded placement co-ordinates of the tex line reference and associated filename, for the reverse call (open file.tex at line I)

Thus SumatraPDF on a fast machine may not get the slower unpacked co-ordinates from a GZ file rather than the uncompressed version, however without knowing the method I also expect that would potentially take longer on any page due to a bigger GZ needing more time to unpack.

For checking if gz is working I suggest first try a small example such my very small (and poor example) highlight.zip at https://github.com/sumatrapdfreader/sumatrapdf/issues/2642#issuecomment-1142729715 

Unpacking the zip as a folder of three files should never respond in the pdf with "No synchronization file found", that would suggest there is a file system issue (such as not using the folder on conventional local drive)?

There will be only two results one of which is an editor **should report "file not found"** from the highlighted function, (it should attempt to open a file not in that folder) the other result should be click on title (or many other text area locations and the tex opens at line 23.
 **OR** gz problem says "No synchronization info at this position" (which can be normal for margins or other white spaces)

**Reply by honekamp** [reply]
I tried the "highlite" example using 3.4.3 and I get very inconsistent results from it. If I start at the TeX file with a forward search while the PDF file is not already open in Sumatra PDF, the PDF opens in Sumatra PDF and the I get the error message "Synchronization file cannot be opened". If I then start an inverse search, the error message is "No synchronization info at this position".

If I load the file into Sumatra PDF and then do an inverse search, everything works like it should. If, after the successful inverse search, I do a forward search, I get the error message "Unknown source file".

**Reply by GitHubRulesOK** [reply]
"unknown source file" (or similar) message from your editor when clicking places other than Title should **be indicating reverse IS working** (that is the point of that highlight test) so confused by you saying getting similar in a forward search
![image](https://user-images.githubusercontent.com/12754913/171665808-864f9536-83fe-456e-965f-7deaa10a223c.png)


**The easier test is double click `Title`** should show line 23 in editor confirming current.tex %file location and %line number were correct (synctex is not top to bottom nor based on word locations) That only proves a known compressed file thats quick to decompress is functioning.

**More telling is**
>not already open in Sumatra PDF, the PDF opens in Sumatra PDF and the I get the error message "Synchronization file cannot be opened"

1st rule of Tex compilation is keep every resource in the one working folder (no /sub/chapter or /aux/folders), its tough enough getting remote or relative TextMF folders to work consistently. so SumatraPDF should know the .synctex.gz is in the same working folder.

**Hinting that SumatraPDF did not detect the companion file from its presence (should not be the case) or had some problem opening the file to decompress**

**Comment by kjk** [comment]
So far what I can tell from logs is that `.synctex.gz` file exists and we call `synctex_scanner_new_with_output_file` which fails for some reason.

This is unfortunately synctex parser code (not written by us) so no idea what's going on there.

I'll add code to pipe synctex parser logs to our logging so that we can get insight on why synctex parser fails.

However, I didn't change synctex parser code in 3.4 so if it worked in 3.3 it should work in 3.4

Given the very large size of `synctex.gz` file I do suspect timing issue i.e. the file not being fully generated when we try to access it.

Maybe a simple hack of re-doing syntex.gz parsing with 1 sec delay (if the file exists and fails the first time) would be a fix.

**Reply by GitHubRulesOK** [reply]
@kjk 
Frustrating or what ? I just retried double click title with my lightweight sample, there was a one second delay, and like OP ***get message and cannot find a log !*** In exactly same session, use the unpacked synctex and for %f %l %c response is `"C:\Users\WDAGUtilityAccount\Desktop\SandBox\LATeX and Reverse-Search\highlite\highlite.tex" 23 0`

<strike>Scratching my head for any co-ordinate reason I wonder if (🎏 red herring alert) there is any relationship to screen setting is 125 dpi?</strike> No otherwise the uncompressed would also not work!

What I did observe was that when removing uncompressed (thus only compressed.gz in folder), initially without refresh it behaved (memory cache?) but as it is normal to refresh pdf, reloading by pressing R resulted in back to error condition!

Ahhhh 😱 I wonder ?? If Sumatra is not writing a temporary LOG ! Where does it think the unpacked GZ data may be  found?, Is issue related to file system IO? Coincidentally, on request debug symbols, I was getting with 1504# "Failed to download symbols." however that currently worked in 15056!
 
![image](https://user-images.githubusercontent.com/12754913/171611885-251b3c27-bf12-418c-906f-bb424a1d8296.png)


Very interesting extreme edge case glitch (so not opening fresh issue), but just to be-aware, when double click is ***exactly*** the position on page where the red circle will later appear the message is automatically instantly dismissed (before time to recognize it), thus could be thought not working !!
![image](https://user-images.githubusercontent.com/12754913/171599577-e41973ad-56de-40e7-9947-b77010b52fee.png)

**Reply by GitHubRulesOK** [reply]
@kjk 
see above HOWEVER after running round in circles with various versions It seems Portable Works for me (always preferred it for LaTeX)
So wonder if its related to all users installation ?? (which is the one that fails)
![image](https://user-images.githubusercontent.com/12754913/171675708-6cc9a354-5879-46e3-9965-a1b302b09596.png)

**Reply by GitHubRulesOK** [reply]
@honekamp 
are you always testing with installed version (you said 3.4.3) but where is sumatrapdf.exe located ?

**Reply by honekamp** [reply]
The exe is located in the default install location, i.e. (if memory serves, currently afk) in c:\Programs\SumatraPDF

**Reply by GitHubRulesOK** [reply]
@honekamp 
Can you please try with a portable copy of 3.4.3 using your files in your installation (its a suggested workaround for now) 

- Run the portable copy once to generate in its own folder sumatrapdf-settings.txt
- Copy from old settings advanced options the line starting `InverseSearchCmdLine = "` to a similar position in new portable settings file
- Then ONLY using your biggest pdf with a synctex.gz check if a double-click works

If you wish you can (backup editor settings) point existing editor at portable sumatrapdf.exe
 if all is well it should allow resumption of 2 way working

**Reply by honekamp** [reply]
@GitHubRulesOK, I followed your recipe, but was unable to get the forward or inverse search to work. I got the already known error messages.

Then I took the portable version from the website and gave it a try. With the portable version everything works just fine, forward and inverse.

BTW, the install directory of Sumatra PDF is in %appdata%\local. My recollection from memory was incorrect.

**Comment by kjk** [comment]
@honekamp I narrowed it down to synctex using zlib's gzopen to try to read the file and failing with rather generic "bad file" error code.

Can you e-mail me (kkowalczyk@gmail.com) the synctex.gz file that's failing?

There's no more logging I can think of that would give me more information.

For some reason synctex is failing to open that particular .gz file. But it works for other .gz files so it's not just generic problem.

If I can step through the code in the debugger, I might learn why this is happening.

At some point I switched from zlib to zlib-ng, which is supposed to be faster. Not sure when but if I did it for 3.4 and zlib-ng has problems with this file, then it would explain why it always worked before 3.4 and now it sometimes fails.

**Comment by kjk** [comment]
@honekamp I got your file, thanks.

So far I was unable to reproduce the problem.

Here's what I did: I created a PDF with a name matching the synctex.gz file, opened it in Sumatra and double clicked in the PDF. The inverse search works i.e. an editor is launched trying to load the original .tex file (which I don't have but that's not important).

I tried with installed (just released) 3.4.4 64 bit, I tried 64 bit / 32 bit, release / debug build.  `.synctex.gz` works in all of them.

When you do similar steps on your machine, does double-clicking not result in inverse search? If yes, then I could only think about differences in processor that make zlib code not work.

So there goes my theory that this is zlib-ng problem.

And we're back to theory of a timing issue i.e. that when you ask Sumatra to do the commands, the `.synctex.gz` is not yet fully created and fails to load.

But that's not something I can recreate easily on my machine.

Do you have additional anti-virus software installed? If yes can you disable it and retry?
Also disable Window's anti-virus software (`Virus & threat protection settings` turn off `Real-time protection`).
Or exclude the relevant folders from virus scanning (it's `Exclusions` for Windows and I'm sure other anti-virus software has similar settings).

The reason is that anti-virus software hooks into OS file operations and scans newly created files. That can make those files unavailable for other processes for a while. The bigger the file, the longer the scanning takes.

This would explain why the file would file to open at one time but then could be opened later.

Also, is the drive where `.synctex.gz` file generated a slow one, like network drive or Dropbox/One Drive or USB drive etc.? Slow drive would increase probability of timing issues.

If yes, try with C:\ drive.

**Reply by honekamp** [reply]
@kjk I can‘t deactivate the virus protection, sorry. The projects I work on are stored on the internal SSD of my computer, so I would expect that speed should not be an issue.

What I don‘t understand about the whole issue is that I was able to get along with my work environment for years, until I installed 3.4.3. And as I mentioned in one other comment above, I can use the portable 3.4.3 and search works fine in both directions. If I go back to 3.3.3 for the installed version, search works again fine in both directions on the same projects that don’t work on 3.4.3.

As I mentioned, I could work around the issue by making sure the the build toolchain for my projects creates non-zipped syntex files. That way, I can at least continue with my work.

I am really grateful for the way you guys have responded to my problem report, it is very much appreciated. But maybe this is the point where it does not make sense to continue investigating without having further clues. I will think about how I can provide more information that may help to dig deeper.

**Comment by kjk** [comment]
@honekamp Makes sense, I ran out of ideas to test as well. Thanks for providing the information.

**Reply by GitHubRulesOK** [reply]
@kjk the only 1 constant is 
that it does not appear to work for some users with 64bit installed SumatraPDF 3.4+
Installed 32bit is OK, either portable is OK

Hardware wise me and OP both seem to be using fast SSDs with i7 (me Dell Win11, HoneKamp Lenovo) so 1 second delay should have discounted too fast on open. No AV other than standard Windows protection.

Now although mostly testing installed within a sandbox I have tried to ensure IF it fails with install inside SB, I also test that it also fails installed outside, the files are local, and I am admin for all (always 1) users with **previewer active in all cases.** 

**Thus PDF is in focus, iFilter is varied e.g. ON inside SandBox, OFF elsewhere. However Pre-viewer is usually active in BOTH
Fail is more often inside SandBox but sometimes outside ! the two environments and their exes should be totally independent however the data files are DOUBLE ACCESS thus one process can interfere with the other ??** and pre-viewer holding files open often causes me issues during install.

My test file is deliberately small (and old) with only two dubious actions, which all standard editors would have handled in a similar fashion, MSNotePad or Write have no redirection but any conforming editor should refocus at line 23, so DoubleClick first PDF Title word opens last lines of TeX, alternatively part highlighted function attempts to open a support file that is not in the work folder (should not exist) very simple, standard.synctex.gz with known odd behaviors.

**Reply by honekamp** [reply]
I have tested 3.4.4 (32 Bit, everything works) vs. 3.4.4 (64 bit, nothing works). Both candidates were non-portable install images. This gives me an uncomfortable flash-back to the one occasion where I was trying to load a 32-bit DLL into a 64-bit executable and my tool failed to execute the DLL code ...

**Reply by GitHubRulesOK** [reply]
@honekamp 
Thanks for feedback
I am running many "dirty" systems (as elevated user) so often cause my own problems, on this occasion It seems to be 64bit inside a Sandbox with Windows PDF search active is possibly having more problems here than without search interfering. However that should not be version related.

One major issue is when users have background file monitors like One Drive, AntiVirus or "Everything" file application etc. Those see the file changes such as unpacking as something to hang on to for inspection. The "One Second" delay should workaround that if they don't hold the file{s} open longer.

**Reply by GitHubRulesOK** [reply]
@honekamp Are you installing pre-viewer / search filter in all 64bit install ?

**Reply by honekamp** [reply]
@GitHubRulesOK, I‘m sorry, I don‘t understand the question.

**Reply by GitHubRulesOK** [reply]
@honekamp 
During your install did you select options
![image](https://user-images.githubusercontent.com/12754913/171910109-6929590b-a588-413b-be24-dbc47cb494ad.png)

**Reply by honekamp** [reply]
No, I have never touched these options before. I always do the straight install with no options.

**Comment by kjk** [comment]
@honekamp the 32-bit vs 64-bit would be consistent with issue being `zlib-ng`. For speed it has assembly implementations that are different for different processor.

I've switched pre-release from zlib-ng back to regular zlib so could you try latest pre-release (both 64-bit and 32-bit release install)?

Another thing I can try is to move reading of `.synctex.gz` from synctex parser code to my code. But this is a bigger change that I really want to make to synctex parser.

**Reply by GitHubRulesOK** [reply]
@kjk 
for me updating to 3.4.4 x64 bit **outside sandbox seems to be good,** no problem with small test file
However now seeing something really **odd** with a link when using pre-rel 3.5.15087 inside sandbox
![image](https://user-images.githubusercontent.com/12754913/171969392-95c9fbb2-b0cc-4a1c-b404-6e6294f66d58.png)

OK that seems to be there is a bleedthrough of **start screen** in the canvas the link is where the about box was before opening the pdf
![image](https://user-images.githubusercontent.com/12754913/171969973-7b6f34cc-ff1d-45ab-9cfd-753708ce3f86.png)

**Reply by kjk** [reply]
Yes, it seems like it. Opened https://github.com/sumatrapdfreader/sumatrapdf/issues/2664 to track

**Reply by GitHubRulesOK** [reply]
@kjk 
just to update that a **fresh install of 3.4.4** inside sandbox appears to be working with synctex (aswell as outside), 
but I was careful NOT to include iFilter, so unconfirmed if that was my factor.

**Reply by honekamp** [reply]
> I've switched pre-release from zlib-ng back to regular zlib so could you try latest pre-release (both 64-bit and 32-bit release install)?

Sorry, @kjk, trying to download the 64 bit pre-release gives me a 404. Unfortunately, I'll not be able to run further tests in the coming weeks. I'll check back as soon as I can and keep you posted on the results I'll get from pre-release-builds.

**Reply by GitHubRulesOK** [reply]
try https://kjkpub.s3.amazonaws.com/software/sumatrapdf/prerel/15093/SumatraPDF-prerel-64-install.exe

**Reply by GitHubRulesOK** [reply]
@kjk since no 32bit pre-rel the 64bit is also 404

**Reply by kjk** [reply]
@GitHubRulesOK Thanks, 404 should be fixed now.

**Comment by honekamp** [comment]
I tried the pre-release from AWS, and unfortunately there is no progress. Forward-search gives me "No synchronization file found" (I proved the existence of the sync file later by installing 3.3.3, which worked on the existing synctex.gz file) and inverse search says "No synchronization info at this position". As mentioned, I'll unfortunately not have time for further test, but will pick up again as soon as possible.

**Comment by T-F-S** [comment]
I followed the discussion and I had the same problem, i.e. synctex search was broken with version 3.4.x, and I had to roll back to version 3.3.3.

Today, I tried 3.4.6 (64 bit installer) and synctex forward and backward search work now for me as in 3.3.3. Version History does not name the problem or its solution, but now it seems to be solved (?).

**Comment by honekamp** [comment]
Today, I tried both 3.4.6 and the latest pre-release. Using the zipped synctex file, sync unfortunately does not work in either of them on my machine. I only get the error messages "No synchronization file found" and "No synchronization info at this position", as reported earlier already.

Meanwhile, I have found a viable solution for changing the PDF build such that the synctex file does not get zipped any longer. As mentioned before, the forward and inverse search works for me on the basis of a non-zipped synctex file.

**Comment by T-F-S** [comment]
I have to supplement my last comment: Version 3.4.6 works for me partially for zipped synctex files, meaning that "No synchronization file found" errors appear more or less often. Closing and opening `sumatrapdf` resets such errors, but they appear frequently and are quite annoying for a larger project.

Using `synctex=-1` to work with unzipped synctex files as @honekamp suggested seems to work much better. I switched my settings yesterday and I did not have any error since.

**Reply by MarcusTressl** [reply]
I have the same problem with ***all*** compressed synctex files in all versions of SumatraPDF that I installed after version 3.1.2 (including 3.4.6): If the synctex file is decompressed, forward and inverse search work well. If it is compressed, neither forward nor inverse search works.  Inverse search responds with a "No synchronization file found" message, so I guess at least this problem is not related to the TeX editor. Working with Windows 10, SumatraPDF x64, WinEdt 5.5. 
[Version 3.1.2 of SumatraPDF x64 works with the compressed synctex file (and the decompressed version).]
Attached is a pdf and its compressed synctex file and the log (if that is of any use) where the problem occurs:
[commands.log](https://github.com/sumatrapdfreader/sumatrapdf/files/9035022/commands.log)
[commands.synctex.gz](https://github.com/sumatrapdfreader/sumatrapdf/files/9035011/commands.synctex.gz)
[commands.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/9035010/commands.pdf)

**Comment by ruijiacao** [comment]
I encountered problems similar to those reported by @honekamp. A good work-around I found was to use the pdfsync package.
